src/script/lua_api/l_ipc.cpp

```diff
@@ -0,0 +1,138 @@
+// Minetest
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include "lua_api/l_ipc.h"
+#include "lua_api/l_internal.h"
+#include "common/c_packer.h"
+#include "server.h"
+#include "debug.h"
+#include <chrono>
+
+typedef std::shared_lock<std::shared_mutex> SharedReadLock;
+typedef std::unique_lock<std::shared_mutex> SharedWriteLock;
+
+static inline auto read_pv(lua_State *L, int idx)
+{
+	std::unique_ptr<PackedValue> ret;
+	if (!lua_isnil(L, idx)) {
+		ret.reset(script_pack(L, idx));
+		if (ret->contains_userdata)
+			throw LuaError("Userdata not allowed");
+	}
+	return ret;
+}
+
+int ModApiIPC::l_ipc_get(lua_State *L)
+{
+	auto *store = getGameDef(L)->getModIPCStore();
+
+	auto key = readParam<std::string>(L, 1);
+
+	{
+		SharedReadLock autolock(store->mutex);
+		auto it = store->map.find(key);
+		if (it == store->map.end())
+			lua_pushnil(L);
+		else
+			script_unpack(L, it->second.get());
```

Comment:
This does the `lua_pushnil` or `script_unpack` while still keeping the lock active.

Reply:
fixing this would require a shared_ptr. another thread could free it